File: ios/Classes/SocketIoNativeClientPlugin.swift
import Flutter
import UIKit
import SocketIO

public class SocketIoNativeClientPlugin: NSObject, FlutterPlugin {
    
    private var commandChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private let streamHandler = SocketStreamHandler()
    
    // Socket.IO properties
    private var socketManager: SocketManager?
    private var socket: SocketIOClient?
    private var activeListeners: [String: UUID] = [:]
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SocketIoNativeClientPlugin()
        instance.setupChannels(with: registrar.messenger())
    }
    
    private func setupChannels(with messenger: FlutterBinaryMessenger) {
        commandChannel = FlutterMethodChannel(name: "devarsh/command", binaryMessenger: messenger)
        eventChannel = FlutterEventChannel(name: "devarsh/events", binaryMessenger: messenger)
        
        commandChannel?.setMethodCallHandler(handle)
        eventChannel?.setStreamHandler(streamHandler)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            switch call.method {
            case "connect":
                guard let args = call.arguments as? [String: Any],
                      let urlString = args["url"] as? String else {
                    result(FlutterError(code: "INVALID_URL", message: "URL not provided or invalid", details: nil))
                    return
                }
                
                if urlString.isEmpty {
                    result(FlutterError(code: "INVALID_URL", message: "URL cannot be empty", details: nil))
                    return
                }
                
                let options = args["options"] as? [String: Any]
                connectSocket(urlString: urlString, options: options, result: result)
                
            case "disconnect":
                disconnectSocket(result: result)
                
            case "emit":
                guard let args = call.arguments as? [String: Any],
                      let event = args["event"] as? String else {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name not provided", details: nil))
                    return
                }
                
                if event.isEmpty {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name cannot be empty", details: nil))
                    return
                }
                
                let data = args["data"]
                emitEvent(event: event, data: data, result: result)
                
            case "listen":
                guard let args = call.arguments as? [String: Any],
                      let event = args["event"] as? String else {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name not provided", details: nil))
                    return
                }
                
                if event.isEmpty {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name cannot be empty", details: nil))
                    return
                }
                
                listenToEvent(event: event, result: result)
                
            case "unlisten":
                guard let args = call.arguments as? [String: Any],
                      let event = args["event"] as? String else {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name not provided", details: nil))
                    return
                }
                
                if event.isEmpty {
                    result(FlutterError(code: "EVENT_ERROR", message: "Event name cannot be empty", details: nil))
                    return
                }
                
                unlistenFromEvent(event: event, result: result)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        } catch {
            result(FlutterError(code: "UNEXPECTED_ERROR", message: "An unexpected error occurred: \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
    
    private func connectSocket(urlString: String, options: [String: Any]?, result: @escaping FlutterResult) {
        do {
            guard let url = URL(string: urlString) else {
                streamHandler.sendEvent(type: "status", payload: ["status": "error", "reason": "Invalid URL format"])
                result(FlutterError(code: "INVALID_URL", message: "The provided URL is invalid", details: urlString))
                return
            }
            
            // Disconnect existing socket if any
            if socket != nil {
                socket?.disconnect()
                socket = nil
                socketManager = nil
                activeListeners.removeAll()
            }
            
            streamHandler.sendEvent(type: "status", payload: ["status": "connecting"])
            
            var config: SocketIOClientConfiguration = []
            var connectParams: [String: Any] = [:]
            
            if let options = options {
                // Handle query and auth parameters
                if let query = options["query"] as? [String: Any] {
                    connectParams.merge(query) { (current, _) in current }
                }
                if let auth = options["auth"] as? [String: Any] {
                    connectParams["auth"] = auth
                }
                if !connectParams.isEmpty {
                    config.insert(.connectParams(connectParams))
                }
                
                // --- Common Properties ---
                if let path = options["path"] as? String { 
                    config.insert(.path(path)) 
                }
                if let reconnect = options["reconnection"] as? Bool { 
                    config.insert(.reconnects(reconnect)) 
                }
                if let attempts = options["reconnectionAttempts"] as? Int { 
                    config.insert(.reconnectAttempts(attempts)) 
                }
                if let secure = options["secure"] as? Bool { 
                    config.insert(.secure(secure)) 
                }
                if let forceNew = options["forceNew"] as? Bool { 
                    config.insert(.forceNew(forceNew)) 
                }
                
                // Time values conversion from milliseconds to seconds
                if let delay = options["reconnectionDelay"] as? Double { 
                    config.insert(.reconnectWait(Int(delay / 1000.0))) 
                }
                if let delayMax = options["reconnectionDelayMax"] as? Double { 
                    config.insert(.reconnectWaitMax(Int(delayMax / 1000.0))) 
                }
                if let jitter = options["randomizationFactor"] as? Double { 
                    config.insert(.randomizationFactor(jitter)) 
                }
                
                // --- iOS-Specific Properties ---
                if let iosConfig = options["extraIOSConfig"] as? [String: Any] {
                    if let log = iosConfig["log"] as? Bool { 
                        config.insert(.log(log)) 
                    }
                    if let headers = iosConfig["extraHeaders"] as? [String: String] { 
                        config.insert(.extraHeaders(headers)) 
                    }
                    if let compress = iosConfig["compress"] as? Bool, compress == true { 
                        config.insert(.compress) 
                    }
                    if let forcePolling = iosConfig["forcePolling"] as? Bool, forcePolling == true { 
                        config.insert(.forcePolling(true)) 
                    }
                    if let forceWebsockets = iosConfig["forceWebsockets"] as? Bool, forceWebsockets == true { 
                        config.insert(.forceWebsockets(true)) 
                    }
                }
            }
            
            socketManager = SocketManager(socketURL: url, config: config)
            socket = socketManager?.defaultSocket
            
            guard let socket = socket else {
                result(FlutterError(code: "CONNECTION_FAILED", message: "Failed to create socket instance", details: nil))
                return
            }
            
            // Set up event handlers
            socket.on(clientEvent: .connect) { [weak self] _, _ in
                let socketId = self?.socket?.sid ?? ""
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "connected", "socketId": socketId])
            }
            
            socket.on(clientEvent: .error) { [weak self] data, _ in
                let errorMessage = String(describing: data.first ?? "Connection error")
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "error", "reason": errorMessage])
            }
            
            socket.on(clientEvent: .disconnect) { [weak self] _, _ in
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "disconnected"])
            }
            
            socket.on(clientEvent: .reconnect) { [weak self] _, _ in
                let socketId = self?.socket?.sid ?? ""
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "connected", "socketId": socketId])
            }
            
            socket.on(clientEvent: .reconnectAttempt) { [weak self] _, _ in
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "connecting"])
            }
            
            // Connect with timeout
            let timeoutValue = options?["timeout"] as? Double ?? 0
            socket.connect(timeoutAfter: timeoutValue > 0 ? timeoutValue / 1000.0 : 0) { [weak self] in
                // This completion handler is called if the connection times out.
                self?.streamHandler.sendEvent(type: "status", payload: ["status": "error", "reason": "Connection timeout"])
            }
            
            result(nil)
            
        } catch {
            streamHandler.sendEvent(type: "status", payload: ["status": "error", "reason": error.localizedDescription])
            result(FlutterError(code: "CONNECTION_FAILED", message: "Connection failed: \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
    
    private func disconnectSocket(result: @escaping FlutterResult) {
        do {
            socket?.disconnect()
            socket = nil
            socketManager = nil
            activeListeners.removeAll()
            result(nil)
        } catch {
            result(FlutterError(code: "DISCONNECTION_FAILED", message: "Failed to disconnect: \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
    
    private func emitEvent(event: String, data: Any?, result: @escaping FlutterResult) {
        do {
            guard let socket = socket, socket.status == .connected else {
                result(FlutterError(code: "NOT_CONNECTED", message: "Socket is not connected", details: nil))
                return
            }
            
            if let data = data as? SocketData {
                socket.emit(event, data)
            } else if data == nil {
                socket.emit(event)
            } else {
               // Convert data to SocketData compatible format
               if let dictData = data as? [String: Any] {
                   socket.emit(event, dictData)
               } else if let arrayData = data as? [Any] {
                   socket.emit(event, arrayData)
               } else if let stringData = data as? String {
                   socket.emit(event, stringData)
               } else if let numberData = data as? NSNumber {
                   // Convert NSNumber to appropriate primitive type
                   if CFNumberIsFloatType(numberData) {
                       socket.emit(event, numberData.doubleValue)
                   } else {
                       socket.emit(event, numberData.intValue)
                   }
               } else if let boolData = data as? Bool {
                   socket.emit(event, boolData)
               } else if let intData = data as? Int {
                   socket.emit(event, intData)
               } else if let doubleData = data as? Double {
                   socket.emit(event, doubleData)
               } else {
                   socket.emit(event, String(describing: data))
               }
            }
            
            result(nil)
        } catch {
            result(FlutterError(code: "EMISSION_FAILED", message: "Failed to emit event '\(event)': \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
    
    private func listenToEvent(event: String, result: @escaping FlutterResult) {
        do {
            guard let socket = socket, socket.status == .connected else {
                result(FlutterError(code: "NOT_CONNECTED", message: "Socket is not connected", details: nil))
                return
            }
            
            let listenerId = socket.on(event) { [weak self] dataArray, ack in
                do {
                    let payload: [String: Any] = ["event": event, "data": dataArray.first ?? NSNull()]
                    self?.streamHandler.sendEvent(type: "socket_event", payload: payload)
                } catch {
                    print("Error processing event '\(event)': \(error.localizedDescription)")
                }
            }
            
            activeListeners[event] = listenerId
            result(nil)
            
        } catch {
            result(FlutterError(code: "EVENT_ERROR", message: "Failed to listen to event '\(event)': \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
    
    private func unlistenFromEvent(event: String, result: @escaping FlutterResult) {
        do {
            if let listenerId = activeListeners[event] {
                socket?.off(id: listenerId)
                activeListeners.removeValue(forKey: event)
            }
            result(nil)
        } catch {
            result(FlutterError(code: "EVENT_ERROR", message: "Failed to unlisten from event '\(event)': \(error.localizedDescription)", details: error.localizedDescription))
        }
    }
}

// MARK: - Stream Handler
class SocketStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    func sendEvent(type: String, payload: Any) {
        guard let sink = self.eventSink else { return }
        do {
            let event: [String: Any] = ["type": type, "payload": payload]
            sink(event)
        } catch {
            print("Error sending event: \(error.localizedDescription)")
        }
    }
} 

File: android/src/main/kotlin/com/devarsh/socket_io_native_client/SocketIoNativeClientPlugin.kt
package com.devarsh.socket_io_native_client

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import java.net.URISyntaxException

/** SocketIoNativeClientPlugin */
class SocketIoNativeClientPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var commandChannel : MethodChannel
    private lateinit var eventChannel : EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var socket: Socket? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        commandChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "devarsh/command")
        commandChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "devarsh/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        try {
            when (call.method) {
                "connect" -> {
                    val url = call.argument<String>("url")
                    if (url.isNullOrEmpty()) {
                        result.error("INVALID_URL", "URL cannot be empty", null)
                        return
                    }

                    val options = call.argument<Map<String, Any>>("options")
                    connectSocket(url, options, result)
                }
                "disconnect" -> {
                    disconnectSocket(result)
                }
                "emit" -> {
                    val event = call.argument<String>("event")
                    val data = call.argument<Any>("data")

                    if (event.isNullOrEmpty()) {
                        result.error("EVENT_ERROR", "Event name cannot be empty", null)
                        return
                    }

                    emitEvent(event, data, result)
                }
                "listen" -> {
                    val event = call.argument<String>("event")

                    if (event.isNullOrEmpty()) {
                        result.error("EVENT_ERROR", "Event name cannot be empty", null)
                        return
                    }

                    listenToEvent(event, result)
                }
                "unlisten" -> {
                    val event = call.argument<String>("event")

                    if (event.isNullOrEmpty()) {
                        result.error("EVENT_ERROR", "Event name cannot be empty", null)
                        return
                    }

                    unlistenFromEvent(event, result)
                }
                else -> {
                    result.notImplemented()
                }
            }
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error in onMethodCall: ${e.message}", e)
            result.error("UNEXPECTED_ERROR", "An unexpected error occurred: ${e.message}", e.toString())
        }
    }

    private fun connectSocket(url: String, options: Map<String, Any>?, result: Result) {
        try {
            sendEvent("status", mapOf("status" to "connecting"))

            // Disconnect existing socket if any
            socket?.disconnect()

            // Validate URL
            try {
                java.net.URI(url)
            } catch (e: URISyntaxException) {
                sendEvent("status", mapOf("status" to "error", "reason" to "Invalid URL format"))
                result.error("INVALID_URL", "Invalid URL format: $url", e.toString())
                return
            }
            val optsBuilder = IO.Options.builder()
            optsBuilder.setUpgrade(true)
            try {
                // --- Common Properties ---
                options?.let { opts ->
                    (opts["reconnection"] as? Boolean)?.let { optsBuilder.setReconnection(it) }
                    (opts["reconnectionAttempts"] as? Number)?.let { optsBuilder.setReconnectionAttempts(it.toInt()) }
                    (opts["reconnectionDelay"] as? Number)?.let { optsBuilder.setReconnectionDelay(it.toLong()) }
                    (opts["reconnectionDelayMax"] as? Number)?.let { optsBuilder.setReconnectionDelayMax(it.toLong()) }
                    (opts["randomizationFactor"] as? Number)?.let { optsBuilder.setRandomizationFactor(it.toDouble()) }
                    (opts["timeout"] as? Number)?.let { optsBuilder.setTimeout(it.toLong()) }
                    (opts["path"] as? String)?.let { optsBuilder.setPath(it) }
                    (opts["forceNew"] as? Boolean)?.let { optsBuilder.setForceNew(it) }
                    (opts["secure"] as? Boolean)?.let { optsBuilder.setSecure(it) }
                    (opts["auth"] as? Map<String, String>)?.let { optsBuilder.setAuth(it) }
                    (opts["query"] as? String)?.let { optsBuilder.setQuery(it) }
                    val transportSet = (options.get("transports") as? List<String>)
                        ?.map { it.toString() }
                        ?.toMutableSet() ?: mutableSetOf()
                    transportSet.add("websocket")
                    optsBuilder.setTransports(transportSet.toTypedArray())
                    // --- Android-Specific Properties ---
                    (options.get("androidConfig") as? Map<String, Any>)?.let { androidConfig ->
                        (androidConfig.get("setExtraHeaders") as Map<String,List<String>>)?.let { optsBuilder.setExtraHeaders(it) }
                    }
                }

                if(socket != null){
                    socket?.disconnect()
                    socket = null
                }

                val opts = optsBuilder.build()
                socket = IO.socket(url, opts)

                socket?.let { sock ->
                    // Connection event handlers
                    sock.on(Socket.EVENT_CONNECT) {
                        val statusData = mapOf("status" to "connected", "socketId" to sock.id())
                        sendEvent("status", statusData)
                    }

                    sock.on(Socket.EVENT_CONNECT_ERROR) { args ->
                        val errorMessage = if (args.isNotEmpty()) args[0].toString() else "Connection failed"
                        Log.e("SocketIoNativeClientPlugin", "Connection error: $errorMessage")
                        sendEvent("status", mapOf("status" to "error", "reason" to errorMessage))
                    }

                    sock.on(Socket.EVENT_DISCONNECT) {
                        sendEvent("status", mapOf("status" to "disconnected"))
                    }

                    sock.connect()
                    result.success(null)
                } ?: run {
                    result.error("CONNECTION_FAILED", "Failed to create socket instance", null)
                }

            } catch (e: Exception) {
                Log.e("SocketIoNativeClientPlugin", "Error configuring socket: ${e.message}", e)
                sendEvent("status", mapOf("status" to "error", "reason" to e.message))
                result.error("CONNECTION_FAILED", "Failed to configure socket: ${e.message}", e.toString())
            }

        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error in connectSocket: ${e.message}", e)
            sendEvent("status", mapOf("status" to "error", "reason" to e.message))
            result.error("CONNECTION_FAILED", "Connection failed: ${e.message}", e.toString())
        }
    }

    private fun disconnectSocket(result: Result) {
        try {
            socket?.disconnect()
            socket = null
            result.success(null)
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error disconnecting: ${e.message}", e)
            result.error("DISCONNECTION_FAILED", "Failed to disconnect: ${e.message}", e.toString())
        }
    }

    private fun emitEvent(event: String, data: Any?, result: Result) {
        try {
            val sock = socket
            if (sock == null || !sock.connected()) {
                result.error("NOT_CONNECTED", "Socket is not connected", null)
                return
            }

            when (data) {
                is Map<*, *> -> sock.emit(event, JSONObject(data as Map<String, Any>))
                is List<*> -> sock.emit(event, *data.toTypedArray())
                null -> sock.emit(event)
                else -> sock.emit(event, data)
            }

            result.success(null)
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error emitting event: ${e.message}", e)
            result.error("EMISSION_FAILED", "Failed to emit event '$event': ${e.message}", e.toString())
        }
    }

    private fun listenToEvent(event: String, result: Result) {
        try {
            val sock = socket
            if (sock == null || !sock.connected()) {
                result.error("NOT_CONNECTED", "Socket is not connected", null)
                return
            }

            sock.on(event) { args ->
                try {
                    val data = if (args.isNotEmpty()) args[0] else null
                    val payload = mapOf("event" to event, "data" to data)
                    sendEvent("socket_event", payload)
                } catch (e: Exception) {
                    Log.e("SocketIoNativeClientPlugin", "Error processing event '$event': ${e.message}", e)
                }
            }

            result.success(null)
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error listening to event: ${e.message}", e)
            result.error("EVENT_ERROR", "Failed to listen to event '$event': ${e.message}", e.toString())
        }
    }

    private fun unlistenFromEvent(event: String, result: Result) {
        try {
            socket?.off(event)
            result.success(null)
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error unlistening from event: ${e.message}", e)
            result.error("EVENT_ERROR", "Failed to unlisten from event '$event': ${e.message}", e.toString())
        }
    }

    private fun sendEvent(type: String, payload: Any) {
        try {
            val event = mapOf("type" to type, "payload" to payload)
            // Ensure we're on the main thread for Flutter platform channel communication
            if (Looper.myLooper() == Looper.getMainLooper()) {
                eventSink?.success(event)
            } else {
                mainHandler.post {
                    try {
                        eventSink?.success(event)
                    } catch (e: Exception) {
                        Log.e("SocketIoNativeClientPlugin", "Error sending event on main thread: ${e.message}", e)
                    }
                }
            }
        } catch (e: Exception) {
            Log.e("SocketIoNativeClientPlugin", "Error sending event: ${e.message}", e)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        commandChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        socket?.disconnect()
        socket = null
    }

    // --- EventChannel.StreamHandler methods ---
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}